src/OpenTelemetry.Extensions.Enrichment/TraceEnrichmentProcessor.cs

```diff
@@ -0,0 +1,43 @@
+// <copyright file="TraceEnrichmentProcessor.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace OpenTelemetry.Extensions.Enrichment;
+
+#pragma warning disable CA1812 // Class is instantiated through dependency injection
+internal sealed class TraceEnrichmentProcessor : BaseProcessor<Activity>
+#pragma warning restore CA1812 // Class is instantiated through dependency injection
+{
+    private readonly TraceEnricher[] traceEnrichers;
+
+    public TraceEnrichmentProcessor(IEnumerable<TraceEnricher> traceEnrichers)
+    {
+        this.traceEnrichers = traceEnrichers.ToArray();
+    }
+
+    public override void OnEnd(Activity activity)
+    {
+        var propertyBag = new TraceEnrichmentBag(activity);
+
+        foreach (var enricher in this.traceEnrichers)
```

Comment:
nit: see if it is faster to use array index to iterate instead of for-each.

Reply:
It it the same, because here traceEnrichers is an array, which means both `foreach` and `for` loops will be compiled into the same IL code

BenchmarkDotNet=v0.13.3, OS=Windows 11 (10.0.22621.1265)
Intel Core i9-10900X CPU 3.70GHz, 1 CPU, 20 logical and 10 physical cores
  [Host]     : .NET Framework 4.8.1 (4.8.9139.0), X64 RyuJIT VectorSize=256
  DefaultJob : .NET Framework 4.8.1 (4.8.9139.0), X64 RyuJIT VectorSize=256


|           Method |     Mean |   Error |  StdDev |   Gen0 | Allocated |
|----------------- |---------:|--------:|--------:|-------:|----------:|
| ForEachProcessor | 147.8 ns | 0.49 ns | 0.44 ns | 0.0892 |     562 B |
|     ForProcessor | 147.9 ns | 0.52 ns | 0.46 ns | 0.0892 |     562 B |


```csharp
[MemoryDiagnoser]
public class ProcessorBenchmark
{
    private static readonly BaseProcessor<Activity> ForProcessor1 = new ForProcessor(new List<TraceEnricher>
    {
        new MyTraceEnricher(),
        new MyTraceEnricher2(),
        new MyTraceEnricher3(),
        new MyTraceEnricher4(),
        new MyTraceEnricher5(),
    });

    private static readonly BaseProcessor<Activity> ForEachProcessor1 = new ForEachProcessor(new List<TraceEnricher>
    {
        new MyTraceEnricher(),
        new MyTraceEnricher2(),
        new MyTraceEnricher3(),
        new MyTraceEnricher4(),
        new MyTraceEnricher5(),
    });

    [Benchmark]
    public void ForEachProcessor()
    {
        ForEachProcessor1.OnEnd(new Activity("test"));
    }

    [Benchmark]
    public void ForProcessor()
    {
        ForProcessor1.OnEnd(new Activity("test"));
    }
}
```